nixos/modules/system/boot/systemd/repart.nix

```diff
@@ -98,16 +113,31 @@ in
             # in the initrd itself.
             ''${config.boot.initrd.systemd.package}/bin/systemd-repart \
                   --definitions=/etc/repart.d \
-                  --dry-run=no
+                  --dry-run=no ${lib.optionalString (initrdCfg.device != null) initrdCfg.device}
             ''
           ];
         };
         # systemd-repart needs to run after /sysroot (or /sysuser, but we don't
         # have it) has been mounted because otherwise it cannot determine the
         # device (i.e disk) to operate on. If you want to run systemd-repart
-        # without /sysroot, you have to explicitly tell it which device to
-        # operate on.
-        after = [ "sysroot.mount" ];
+        # without /sysroot (i.e. to create the root partition), you have to
+        # explicitly tell it which device to operate on. The service then needs
+        # to be ordered to run after this device is available.
+        unitConfig =
+          let
+            # Imitates `systemd-escape --path` but will not work correctly for
+            # corner cases.
+            escapePath = path: lib.replaceStrings [ "/" ] [ "-" ] (lib.removePrefix "/" path);
```

Comment:
We have an escape-path utility in systemd-lib.nix

Reply:
Do you mean [mkPathSafeName](https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/systemd-lib.nix#L13)? I thought about using it but it doesn't really do what I need (i.e strip the first `/` and change `/` in the middle to `-`). Is it supposed to mimic `systemd-escape --path`? Then maybe it would make sense to change it. However I don't really know the blast radius of messing up this escape function.

Reply:
@nikstur He means `escapeSystemdPath`, right?

Reply:
Awesome, I just looked in the wrong places.